README
NAME
    B::Stats - print optree statistics

SYNOPSIS
      perl -MB::Stats myprog.pl # all
      perl -MO=Stats myprog.pl  # compile-time only
      perl -MB::Stats[,OPTIONS] myprog.pl

DESCRIPTION
    Print statistics for all generated ops.

    static analysis at compile-time, static analysis at end-time to include
    all runtime added modules, and dynamic analysis at run-time, as with a
    profiler.

    The purpose is to help you in your goal:

        no bloat;

OPTIONS
    Options can be bundled: "-c,-e,-r" eq "-cer"

    -c *static*
        Do static analysis at compile-time. This does not include all
        run-time require packages. Invocation via -MO=Stats does this
        automatically.

    -e *end*
        Do static analysis at end-time. This is includes all run-time
        require packages. This calculates the heap space for the optree.

    -r *run*
        Do dynamic run-time analysis of all actually visited ops, similar to
        a profiler. Single ops can be called multiple times.

        Warning: Since perl v5.31.9 the -r option requires Opcodes to be
        pre-loaded. If this is not possible, it does not work anymore and is
        disabled.

    -a *all (default)*
        Same as -cer: static compile-time, end-time and dynamic run-time.

    -t *table*
        Tabular list of -c, -e and -r results.

    -u *summary*
        Short summary only, no op class. With -t only the final table(s).

    -F *Files*
        Prints included file names

    -x *fragmentation* NOT YET
        Calculates the optree *fragmentation*. 0.0 is perfect, 1.0 is very
        bad.

        A perfect optree has no null ops and every op->next is immediately
        next to the op.

    -f<op,...> *filter* NOT YET
        Filter for op names and classes. Only calculate the given ops, resp.
        op class.

          perl -MB::Stats,-fLOGOP,-fCOP,-fconcat myprog.pl

    -llogfile
        Print output only to this file. Default: STDERR

          perl -MB::Stats,-llog myprog.pl

METHODS
    compile
        Static -c check at CHECK time. Triggered by -MO=Stats,-OPTS

    rcount ($opcode)
        Returns run-time count per op type.

    rcount_all()
        Returns an AV ref for all opcounts, indexed by opcode.

    reset_rcount()
        Resets to opcount array for all ops to 0

    output ($count-hash, $ops, $name)
        General formatter

    output_runtime
        -r formatter.

        Prepares count hash from the runtime generated structure in XS and
        calls output().

    output_table
        -t formatter

        Does not work together with -r

LICENSE
    This module is available under the same licences as perl, the Artistic
    license and the GPL.

SEE ALSO


Changes
-*- change-log -*-

0.10	2020-02-26 rurban
	* disable 'require Opcodes' with -r, the dynamic run-time stats
	  since 5.31.9 because of problems in the END phase. RT #131963

0.09	2014-08-08 rurban
	* fix Stats.xs for 5.6. no -Dv on 5.6, combine with -Dt instead
	* fix c_minus for 5.6. Needed to seperate $t Dump.
	* add missing Changes to MANIFEST

0.08	2014-06-06 rurban
	* fix overhead subtraction, and count OPCLASS at the end
	  just -E0 is still wrong, i.e. still gives the -E overhead
	* initialize empty B::Stats::Minus before overhead calc
	* change -fOP to -fBASEOP, -fPVOP to -fPVOP_OR_SVOP for consistency
	* add 3 missing op classes BASEOP_OR_UNOP FILESTATOP LOOPEXOP
	  (thanks bulk88 for issue #5)
	* print error on -t,-r, i.e. without -c and -e stats

0.07	2014-05-19 rurban
	* fix runtime stats (-r) to be meaningful/accurate, fixed various bugs (bulk88)
	* fixed initially computed B::Stats::Minus (windows quoting issue)
	* document new functions rcount_all() and reset_rcount()

0.06	2014-05-13 rurban
	* Win32 fixes by bulk88
	* stabilize B::Stats::Minus creation
	* replace each in walksymtable by foreach

0.05	2012-05-02 rurban
	* protect against empty B::Stats::Minus::overhead key warnings,
	when more op types are loaded at END than at COMPILE time.

0.04	2011-12-12 rurban
	* use run-time not begin-time %INC on -r or -e for files and lines.
	* fixed empty END runops loops caused by format
	* implemented -l<logfile> and -f<filter>
	* check for 2x import()'ed

0.03	2011-12-12 rurban
	* forgot to add c_minus and lib/B/Stats/Minus.pm to MANIFEST

0.02	2011-12-09 rurban
	* added generated B::Stats::Minus to substract our own overhead
	* implemented switch bundling
	* added t/exact.t to test B::Stats::Minus

0.01	2011-12-06 rurban
	* First release as B-Stats-0.01


t/t.pl
sub run {
    if($ENV{H}) {
    print "y\n";
    } else {
    print "n\n";
    }
}
run();


Makefile.PL
use strict;
use warnings;

use ExtUtils::MakeMaker 6.17; # perl-5.6.2

my %WriteMakefileArgs =
  (
   "NAME" 		=> "B::Stats",
   "VERSION_FROM" 	=> "lib/B/Stats.pm",
   "ABSTRACT"		=> "print optree statistics",
   "AUTHOR" 		=> "Reini Urban <rurban\@cpan.org>",
   "LICENSE" 		=> "perl",
   "PREREQ_PM" =>  {
		    'Opcodes' 	=> "0.11",
		    'B' 	=> 0,	# 5.6.2 has no version, 5.8.1 has 1.02
		    'Data::Dumper' => 0,# in core since perl 5.005
		   },
   "SIGN" => 1,
   ($ExtUtils::MakeMaker::VERSION gt '6.46' ?
    ('META_MERGE'  =>
     {
      resources =>
      {
	license     => 'http://dev.perl.org/licenses/',
	bugtracker  => 'RT',
	repository  => 'git://github.com/rurban/B-Stats.git',
      }}) : ()),
  );

WriteMakefile(%WriteMakefileArgs);

sub MY::depend {
  "
lib/B/Stats/Minus.pm : Makefile c_minus
	\$(TOUCH) \$@
	\$(MAKE) pm_to_blib
	\$(FULLPERL) c_minus

README : lib/B/Stats.pm
	pod2text lib/B/Stats.pm > README

release : test dist
	-git tag -f \$(VERSION)
	cpan-upload \$(DISTVNAME).tar\$(SUFFIX)
	git push
	git push --tags
";
}

# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 100
# End:
# vim: expandtab shiftwidth=4:


lib/B/Stats/Minus.pm
package B::Stats::Minus;
# Automatically generated by B::Stats Makefile.PL
# Install-time generated data, to calculate the B::Stats overhead.
#   c includes B with its 14 dependencies
#   e adds Opcodes
#   UPPERCASE for -E (features overhead)
$overhead = {
              'e' => {},
              'c' => {
                       'BEGIN' => '0',
                       'Can\'t' => '0',
                       'Compilation' => '0'
                     },
              'r' => {},
              'C' => {
                       'BEGIN' => '0',
                       'Can\'t' => '0',
                       'Compilation' => '0'
                     },
              'R' => {},
              'E' => {}
            };
$test = {
          'nextstate' => 1,
          'null' => 1,
          'enter' => 1,
          'leave' => 1
        };
$test_E = {
            'bit_and' => 2,
            'next' => 2,
            'pushmark' => 15,
            'require' => 2,
            'sassign' => 9,
            'not' => 1,
            'substr' => 6,
            'lineseq' => 7,
            'substcont' => 2,
            'or' => 6,
            'padsv' => 32,
            'leavesub' => 5,
            'seq' => 4,
            'eq' => 1,
            'scope' => 7,
            'values' => 1,
            'cond_expr' => 1,
            'enter' => 7,
            'null' => 69,
            'subst' => 2,
            'const' => 22,
            'bit_or' => 1,
            'entersub' => 9,
            'shift' => 6,
            'delete' => 2,
            'nextstate' => 39,
            'and' => 7,
            'padhv' => 11,
            'leave' => 7,
            'rv2hv' => 3,
            'gv' => 19,
            'method_named' => 1,
            'gvsv' => 12,
            'rv2av' => 10,
            'unshift' => 2,
            'sprintf' => 2,
            'unstack' => 2,
            'helem' => 5,
            'leaveloop' => 2,
            'enterloop' => 2,
            'concat' => 4,
            'complement' => 2,
            'return' => 1,
            'exists' => 6
          };
# $test for -e1
# $test_E for -E1
1;


lib/B/Stats.pm
package B::Stats;
use strict;
our $VERSION = '0.10';

# TODO
# exact: probably use Opcodes and DynaLoader at BEGIN for c_minus.
# less overhead: do more in XS: _class, _count_op, _walkoptree_simple, _walksymtable
# detect -E usage, features overhead

=head1 NAME

B::Stats - print optree statistics

=head1 SYNOPSIS

  perl -MB::Stats myprog.pl # all
  perl -MO=Stats myprog.pl  # compile-time only
  perl -MB::Stats[,OPTIONS] myprog.pl

=head1 DESCRIPTION

Print statistics for all generated ops.

static analysis at compile-time,
static analysis at end-time to include all runtime added modules,
and dynamic analysis at run-time, as with a profiler.

The purpose is to help you in your goal:

    no bloat;

=head1 OPTIONS

Options can be bundled: C<-c,-e,-r> eq C<-cer>

=over

=item -c I<static>

Do static analysis at compile-time. This does not include all run-time require packages.
Invocation via -MO=Stats does this automatically.

=item -e I<end>

Do static analysis at end-time. This is includes all run-time require packages.
This calculates the heap space for the optree.

=item -r I<run>

Do dynamic run-time analysis of all actually visited ops, similar to a profiler.
Single ops can be called multiple times.

Warning: Since perl v5.31.9 the -r option requires Opcodes to be pre-loaded.
If this is not possible, it does not work anymore and is disabled.

=item -a I<all (default)>

Same as -cer: static compile-time, end-time and dynamic run-time.

=item -t I<table>

Tabular list of -c, -e and -r results.

=item -u I<summary>

Short summary only, no op class.
With -t only the final table(s).

=item -F I<Files>

Prints included file names

=item -x I<fragmentation>  B<NOT YET>

Calculates the optree I<fragmentation>. 0.0 is perfect, 1.0 is very bad.

A perfect optree has no null ops and every op->next is immediately next
to the op.

=item -f<op,...> I<filter>  B<NOT YET>

Filter for op names and classes. Only calculate the given ops, resp. op class.

  perl -MB::Stats,-fLOGOP,-fCOP,-fconcat myprog.pl

=item -lF<logfile>

Print output only to this file. Default: STDERR

  perl -MB::Stats,-llog myprog.pl

=back

=head1 METHODS

=over

=cut

our (%B_inc, %B_env);
BEGIN { %B_inc = %INC; }

# B includes 14 files and 3821 lines. overhead subtracted with B::Stats::Minus
use B;
use B::Stats::Minus;
# XSLoader adds 0 files and 0 lines, already with B.
# Changed to DynaLoader
# Opcodes-0.10 adds 6 files and 5303-3821 lines: Carp, AutoLoader, subs
# Opcodes-0.11 adds 2 files and 4141-3821 lines: subs
# use Opcodes; # deferred to run-time below
our ($static, @runtime, $compiled, $imported, $LOG);
my (%opt, $nops, $rops, @all_subs, $frag, %roots);
my ($c_count, $e_count, $r_count);

# check options
sub import {
  $DB::single = 1 if defined &DB::DB;
#print STDERR "opt: ",join(',',@_),"; "; # for Debugging
  for (@_) { # switch bundling without Getopt bloat
    if (/^-?([acerxtFu])(.*)$/) {
      $opt{$1} = 1;
      my $rest = $2;
      do {
	if ($rest =~ /^-?([acerxtFu])(.*)$/) {
	  $opt{$1} = 1;
	  $rest = $2;
	}
      } while $rest;
    }
    # taking multiple arguments: -ffilter
    if (/^-?f(.*)$/) {
      my $arg = $1;
      if ($arg =~ /^[A-Z_]*OP$/) {
        my @optype = qw(BASEOP UNOP BINOP LOGOP LISTOP PMOP SVOP PADOP PVOP_OR_SVOP
                        LOOP COP BASEOP_OR_UNOP FILESTATOP LOOPEXOP);
	$opt{f}->{class}->{$arg} = 1;
	die "invalid B::Stats,-fOPCLASS argument: $arg\n"
	  unless grep {$arg eq $_} @optype;
	# pre-expand names for the class
  # RT #131963, too late to reload Opcodes
  unless (${^GLOBAL_PHASE} eq 'END' and $] >= 5.031009 and !defined(\&Opcodes::opcodes)) {
    my $maxo = Opcodes::opcodes();
    for my $i (0..$maxo-1) {
      my $name = Opcodes::opname($i);
      my $class = $optype[ Opcodes::opclass($i) ];
      if ($class eq $arg) {
        $opt{f}->{name}->{$name} = 1;
      }
    }
  }
      } elsif ($arg =~ /^[a-z_]+$/) {
	$opt{f}->{name}->{$arg} = 1;
      } else {
	die "invalid B::Stats,-ffilter argument: $arg\n";
      }
    }
    if (/^-?(l)(.*)$/) { # taking arguments: -llogfile
      $opt{$1} = $2;
      open $LOG, ">", $opt{l} or die "Cannot write to B::Stats,-llogfile: $opt{l}\n";
    }
  }

  $opt{a} = 1 if !$opt{c} and !$opt{e} and !$opt{r}; # default
  $opt{c} = $opt{e} = $opt{r} = 1 if $opt{a};
#warn "%opt: ",keys %opt,"\n"; # for Debugging
  $LOG = \*STDERR unless $opt{l};
  $imported = 1;
}

sub _class {
    my $name = ref shift;
    $name =~ s/^.*:://;
    return $name;
}

# static
sub _count_op {
  my $op = shift;
  $nops++; # count also null ops
  if ($$op) {
    $static->{name}->{$op->name}++;
    # $static->{class}->{_class($op)}++;
  }
}

# collect subs and stashes before B is loaded
# XXX not yet used. we rather use B::Stats::Minus
sub _collect_env {
  %B_env = { 'B::Stats' => 1 };
  _xs_collect_env() if $INC{'DynaLoader.pm'};
}

# from B::Utils
our $sub;
sub B::GV::_mypush_starts {
  my $name = $_[0]->STASH->NAME."::".$_[0]->SAFENAME;
  return unless ${$_[0]->CV};
  my $cv = $_[0]->CV;
  if ($cv->PADLIST->can("ARRAY")
      and $cv->PADLIST->ARRAY
      and $cv->PADLIST->ARRAY->can("ARRAY"))
  {
    push @all_subs, $_->ROOT
      for grep { _class($_) eq "CV" } $cv->PADLIST->ARRAY->ARRAY;
  }
  return unless ${$cv->START} and ${$cv->ROOT};
  $roots{$name} = $cv->ROOT;
};
sub B::SPECIAL::_mypush_starts{}

sub _walkops {
  my ($callback, $data) = @_;
  # _collect_env() unless %B_env;
  %roots  = ( '__MAIN__' =>  B::main_root()  );
  _walksymtable(\%main::,
	       '_mypush_starts',
	       sub {
		 return if scalar grep {$_[0] eq $_."::"} ('B::Stats');
		 1;
	       }, # Do not eat our own children!
	       '');
  push @all_subs, $_->ROOT
    for grep { _class($_) eq "CV" } B::main_cv->PADLIST->ARRAY->ARRAY;
  for $sub (keys %roots) {
    _walkoptree_simple($roots{$sub}, $callback, $data);
  }
  # catches __ANON__
  for (@all_subs) {
    _walkoptree_simple($_, $callback, $data);
  }
}

sub _walkoptree_simple {
  my ($op, $callback, $data) = @_;
  $callback->($op,$data);
  if ($$op && ($op->flags & B::OPf_KIDS)) {
    for (my $kid = $op->first; $$kid; $kid = $kid->sibling) {
      _walkoptree_simple($kid, $callback, $data);
    }
  }
}

sub _walksymtable {
    my ($symref, $method, $recurse, $prefix) = @_;
    my ($sym, $ref, $fullname);
    no strict 'refs';
    $prefix = '' unless defined $prefix;
    foreach my $sym (keys %$symref) {
        my $ref = $symref->{$sym};
        $fullname = "*main::".$prefix.$sym;
	if ($sym =~ /::$/) {
	    $sym = $prefix . $sym;
	    if (B::svref_2object(\*$sym)->NAME ne "main::" &&
		$sym ne "<none>::" && &$recurse($sym))
	    {
               _walksymtable(\%$fullname, $method, $recurse, $sym);
	    }
	} else {
           B::svref_2object(\*$fullname)->$method();
	}
    }
}

=item compile

Static -c check at CHECK time. Triggered by -MO=Stats,-OPTS

=cut

sub compile {
  import(@_) unless $imported; # check options via O
  $compiled++;
  $opt{c} = 1;
  return sub {
    $nops = 0;
    _walkops(\&_count_op);
    output($static, $nops, 'static compile-time');
  }
}

=item rcount ($opcode)

Returns run-time count per op type.

=item rcount_all()

Returns an AV ref for all opcounts, indexed by opcode.

=item reset_rcount()

Resets to opcount array for all ops to 0

=item output ($count-hash, $ops, $name)

General formatter

=cut

sub output {
  my ($count, $ops, $name) = @_;

  my %name = (
    'static compile-time' => 'c',
    'static end-time'     => 'e',
    'dynamic run-time'    => 'r'
    );
  my $key = $name{$name};
  my $lines = 0;
  my $inc = $key eq 'c' ? \%B_inc : \%INC;
  my $files = scalar keys %$inc;
  for (values %$inc) {
    print $LOG $_,"\n" if $opt{F};
    open IN, "<", "$_";
    # Todo: skip pod?
    while (<IN>) { chomp; s/#.*//; next if not length $_; $lines++; };
    close IN;
  }
  for (qw(_files _lines _ops)) {
    $B::Stats::Minus::overhead->{$key}{$_} = 0 unless $B::Stats::Minus::overhead->{$key}{$_};
  }
  $files -= $B::Stats::Minus::overhead->{$key}{_files};
  $lines -= $B::Stats::Minus::overhead->{$key}{_lines};
  $ops -= $B::Stats::Minus::overhead->{$key}{_ops};
  print $LOG "\nB::Stats $name:\nfiles=$files\tlines=$lines\tops=$ops\n";
  return if $opt{t} and $opt{u};

  print $LOG "\nop name:\n";
  for (sort { $count->{name}->{$b} <=> $count->{name}->{$a} }
       keys %{$count->{name}}) {
    my $l = length $_;
    my $c = $count->{name}->{$_};
    $c -= $B::Stats::Minus::overhead->{$key}{$_} if exists $B::Stats::Minus::overhead->{$key}{$_};
    $count->{name}->{$_} = $c;
    next if $opt{f} and !$opt{f}->{name}->{$_};
    next if !$c;
    print $LOG $_, " " x (10-$l), "\t", $c, "\n";
  }
  # FIXME: no OPCLASS overhead substraction
  unless ($opt{u}) {
    print $LOG "\nop class:\n";

    #warn "before \$INC{Opcodes.pm}: ",$INC{'Opcodes.pm'}, ${^GLOBAL_PHASE};
    # RT #131963, too late to reload Opcodes
    if ($] >= 5.031009 && ${^GLOBAL_PHASE} eq 'END') {
      return $count;
    }
    require Opcodes;
    #warn "after \$INC{Opcodes.pm}: ",$INC{'Opcodes.pm'};
    my $maxo = Opcodes::opcodes();
    my @optype = qw(BASEOP UNOP BINOP LOGOP LISTOP PMOP SVOP PADOP PVOP_OR_SVOP
                    LOOP COP BASEOP_OR_UNOP FILESTATOP LOOPEXOP);
    for my $i (0..$maxo-1) {
      my $name = Opcodes::opname($i);
      if ($name) {
        my $class = $optype[ Opcodes::opclass($i) ];
        next if $opt{f} and !$opt{f}->{name}->{$name};
        next if $opt{f} and !$opt{f}->{class}->{$class};
        $count->{class}->{ $class } += $count->{name}->{ $name };
      } else {
        warn "invalid name for opcount[$i]";
      }
    }
    for (sort { $count->{class}->{$b} <=> $count->{class}->{$a} }
	 keys %{$count->{class}}) {
      next if $opt{f} and !$opt{f}->{class}->{$_};
      next if !$count->{class}->{$_};
      my $l = length $_;
      print $LOG $_, " " x (10-$l), "\t", $count->{class}->{$_}, "\n";
    }
  }
  $count;
}

=item output_runtime

-r formatter.

Prepares count hash from the runtime generated structure in XS and calls output().

=cut

sub output_runtime {
  $r_count = {};
  my $r_countarr = $_[0];

  #require DynaLoader;
  #our @ISA = ('DynaLoader');
  #DynaLoader::bootstrap('B::Stats', $VERSION);
  # RT #131963, too late to load Opcodes
  if ($] >= 5.031009 && ${^GLOBAL_PHASE} eq 'END' && !defined(\&Opcodes::opcodes)) {
    return;
  }
  require Opcodes;
  my $maxo = Opcodes::opcodes();
  # @optype only since 5.8.9 in B
  my @optype = qw(BASEOP UNOP BINOP LOGOP LISTOP PMOP SVOP PADOP PVOP_OR_SVOP
                  LOOP COP BASEOP_OR_UNOP FILESTATOP LOOPEXOP);
  for my $i (0..$maxo-1) {
    if (my $count = $r_countarr->[$i]) {
      my $name = Opcodes::opname($i);
      if ($name) {
	my $class = $optype[ Opcodes::opclass($i) ];
	next if $opt{f} and !$opt{f}->{name}->{$name};
	next if $opt{f} and !$opt{f}->{class}->{$class};
	$r_count->{name}->{ $name } += $count;
	$rops += $count;
	# $r_count->{class}->{ $class } += $count;
      } else {
	warn "invalid name for opcount[$i]";
      }
    }
  }
  $r_count = output($r_count, $rops, 'dynamic run-time');
}

=item output_table

-t formatter

Does not work together with -r

=cut

sub _output_tline {
  my $n = shift;
  my $name = $n.(" "x(12-length($n)));
  return if $opt{f} and !$opt{f}->{name}->{$n};
  my ($c, $e, $r) = ($c_count->{name}->{$n},
		     $e_count->{name}->{$n},
		     $r_count->{name}->{$n});
  return if !$c and !$e and !$r;
  print $LOG join("\t", ($name, $c, $e, $r)), "\n";
}

sub output_table {
  # XXX we have empty runops runs with format.
  #my $x = 0;
  #for (keys %{$r_count->{name}}) {
  #  $x++ if $r_count->{name}->{$_};
  #}
  #return unless $x;
  print $LOG "
B::Stats table:
           	-c	-e	-r
";
  warn "Cannot use -t with -r only\n" if !$opt{c} and !$opt{e};
  if ($e_count and %$e_count) {
    for (sort { $e_count->{name}->{$b} <=> $e_count->{name}->{$a} }
         keys %{$e_count->{name}}) {
      _output_tline($_);
    }
  } else {
    for (sort { $c_count->{name}->{$b} <=> $c_count->{name}->{$a} }
         keys %{$c_count->{name}}) {
      _output_tline($_);
    }
  }
}

=back

=cut

# Called not via -MO=Stats, rather -MB::Stats
CHECK {
  compile->() if !$compiled and $opt{c};
}

sub _end { #void _end($refToArrOfRuntimeCounts)
  $c_count = $static;
  if ($opt{e}) {
    $nops = 0;
    $static = {};
    _walkops(\&_count_op);
    $e_count = output($static, $nops, 'static end-time');
  }
  output_runtime($_[0]) if $opt{r};
  if ( $opt{t} and ($] < 5.014 or ${^GLOBAL_PHASE} ne 'DESTRUCT') ) {
    output_table;
  }
}

=head1 LICENSE

This module is available under the same licences as perl, the Artistic
license and the GPL.

=head1 SEE ALSO

=cut

XSLoader::load 'B::Stats', $VERSION;
1;
